src/ngraph_builder1.h

```diff
@@ -0,0 +1,138 @@
+// translation, the heavy-lifting being done by ngraph_translateops.h;
+// It merely does some bookkeeping, such as process the nodes in correct
+// order, keep track of and retrieve parent nGraph nodes etc.
+
+class Builder1 {
+  using VectNg = std::vector<shared_ptr<ng::Node>>;
+  using TranslatorFn = function<Status(
+      const Node*, VectNg&, const std::vector<const Tensor*>&, VectNg&)>;
+
+ public:
+  // Note: in case we want to get rid of Initialize, pass an
+  // OpKernelConstruction* ctx to the constructor, and use OP_REQUIRES_OK
+  // instead of TF_RETURN_IF_ERROR. This gets rid of Initialize() as it
+  // can handle errors in construction.
+  // But in the case of overloaded constructor that does not accept a ctx,
+  // which is used for OpExecuter test class, we cannot handle error during
+  // construction.
+  // Hence keeping the Initialize() function
+
+  // This constructor is for actual use (encapsulate_op)
+  Builder1(const Graph& tf_graph, OpKernelConstruction* ctx)
+      : tf_graph(tf_graph) {}
+
+  // And this version is for tests (OpExecuter)
+  Builder1(const Graph& tf_graph) : Builder1(tf_graph, nullptr) {}
+
+  // TranslateGraph is overloaded. This is for actual use (encapsulate_op)
+  Status TranslateGraph(OpKernelContext* ctx,
+                        std::shared_ptr<ngraph::Function>& ng_function);
+
+  // And this version is for tests (OpExecuter)
+  Status TranslateGraph(const std::vector<TensorShape>&,
+                        const std::vector<const Tensor*>&,
+                        shared_ptr<ng::Function>&);
+
+ private:
+  // The op map holds a mapping from TensorFlow op names (strings) to
+  // vector of generated nGraph nodes. Since we process nodes in toposort
+  // order, it is guaranteed when processing a TF node, its parents
+  // will already have been processed and safely stowed in ng_op_map
+  std::unordered_map<std::string, VectNg> ng_op_map;
+
+  // Prevent Initialize from running twice
+  bool is_initialized = false;
+  const Graph& tf_graph;
+
+  // An array that will be useful in creating the static_input_map
+  std::vector<bool> m_input_is_static;
+
+  // Vectors containing TF nodes in 3 bins: inputs, outputs, actual ops
+  vector<const Node *> tf_params, tf_ret_vals, tf_ops;
+
+  // This map tells us which inputs to read for a particular node. If no
+  // information is present explicitly in the map, we read all inputs
+  // from 0 to num_inputs-1
+  const static std::map<const string, vector<int>> INPUT_INDEX_MAP;
+
+  // A map from Tf op type_string to a TranslateOp
+  const static std::map<const string, Builder1::TranslatorFn> TRANSLATE_OP_MAP;
+
+  // Given a TF node, return its corresponding TranslateOp function and required
+  // input indexes. A wrapper for TRANSLATE_OP_MAP and INPUT_INDEX_MAP
+  Status GetOpTranslationRequirements(const Node*, Builder1::TranslatorFn&,
+                                      vector<int>&);
```

Comment:
I think TF style favors pointers over non-const references (not 100% sure though)

Reply:
https://google.github.io/styleguide/cppguide.html#Reference_Arguments

It seems to prefer a const T&, and never a raw T&. It prefer's T* over T& for 'return' values.

But, our code makes a lot of use of T& for returning. For example TranslateGraph itself uses std::shared_ptr<ngraph::Function>& to return the created ng_function. 

The 'cons' of not using T& as stated in the style guide is:
References can be confusing, as they have value syntax but pointer semantics.

I am not sure if addressing that con is worth the effort of overhauling all places where we have already used T& for returning values